pkg/storage/tsdb/caching_bucket.go

```diff
@@ -36,47 +62,133 @@ func (cfg *ChunksCacheConfig) RegisterFlagsWithPrefix(f *flag.FlagSet, prefix st
 	f.DurationVar(&cfg.SubrangeTTL, prefix+"subrange-ttl", 24*time.Hour, "TTL for caching individual chunks subranges.")
 }
 
-// Validate the config.
 func (cfg *ChunksCacheConfig) Validate() error {
-	if cfg.Backend != "" && cfg.Backend != string(storecache.MemcachedBucketCacheProvider) {
-		return fmt.Errorf("unsupported cache backend: %s", cfg.Backend)
+	return cfg.CacheBackend.Validate()
+}
+
+type MetadataCacheConfig struct {
+	CacheBackend `yaml:",inline"`
+
+	TenantsListTTL         time.Duration `yaml:"tenants_list_ttl"`
+	TenantBlocksListTTL    time.Duration `yaml:"tenant_blocks_list_ttl"`
+	ChunksListTTL          time.Duration `yaml:"chunks_list_ttl"`
+	MetafileExistsTTL      time.Duration `yaml:"metafile_exists_ttl"`
+	MetafileDoesntExistTTL time.Duration `yaml:"metafile_doesnt_exist_ttl"`
+	MetafileContentTTL     time.Duration `yaml:"metafile_content_ttl"`
+	MetafileMaxSize        int           `yaml:"metafile_max_size"`
+}
+
+func (cfg *MetadataCacheConfig) RegisterFlagsWithPrefix(f *flag.FlagSet, prefix string) {
+	f.StringVar(&cfg.Backend, prefix+"backend", "", fmt.Sprintf("Backend for metadata cache, if not empty. Supported values: %s.", CacheBackendMemcached))
+
+	cfg.Memcached.RegisterFlagsWithPrefix(f, prefix+"memcached.")
+
+	f.DurationVar(&cfg.TenantsListTTL, prefix+"tenants-list-ttl", 15*time.Minute, "How long to cache list of tenants in the bucket.")
+	f.DurationVar(&cfg.TenantBlocksListTTL, prefix+"tenant-blocks-list-ttl", 15*time.Minute, "How long to cache list of blocks for each tenant.")
+	f.DurationVar(&cfg.ChunksListTTL, prefix+"chunks-list-ttl", 24*time.Hour, "How long to cache list of chunks for a block.")
+	f.DurationVar(&cfg.MetafileExistsTTL, prefix+"metafile-exists-ttl", 2*time.Hour, "How long to cache information that block metafile exists.")
+	f.DurationVar(&cfg.MetafileDoesntExistTTL, prefix+"metafile-doesnt-exist-ttl", 15*time.Minute, "How long to cache information that block metafile doesn't exist.")
+	f.DurationVar(&cfg.MetafileContentTTL, prefix+"metafile-content-ttl", 24*time.Hour, "How long to cache content of the metafile.")
+	f.IntVar(&cfg.MetafileMaxSize, prefix+"metafile-max-size", 1*1024*1024, "Maximum size of metafile content to cache.")
+}
+
+func (cfg *MetadataCacheConfig) Validate() error {
+	return cfg.CacheBackend.Validate()
+}
+
+func CreateCachingBucket(chunksConfig ChunksCacheConfig, metadataConfig MetadataCacheConfig, bkt objstore.Bucket, logger log.Logger, reg prometheus.Registerer) (objstore.Bucket, error) {
+	cfg := storecache.NewCachingBucketConfig()
+	cachingConfigured := false
+
+	chunksCache, err := createCache("chunks-cache", chunksConfig.Backend, chunksConfig.Memcached, logger, reg)
+	if err != nil {
+		return nil, errors.Wrapf(err, "chunks-cache")
+	}
+	if chunksCache != nil {
+		cachingConfigured = true
+		cfg.CacheGetRange("chunks", chunksCache, isTSDBChunkFile, chunksConfig.SubrangeSize, chunksConfig.ObjectSizeTTL, chunksConfig.SubrangeTTL, chunksConfig.MaxGetRangeRequests)
```

Comment:
We should rename `ObjectSizeTTL` to `AttributesTTL` (or something like this) and rename the YAML/CLI config too.